### PR TITLE
[PERF] Don't use array methods

### PR DESCRIPTION
**Pull request by chadhietala**
We shouldn't use array methods as they allocate and close over items that need to be GC'd. Memory preasure can be an issue and we should do everything to reduce that preasure.

### CODE REVIEW

File: addon/-private/serializers/embedded-records-mixin.js [hunk 1]
```diff
@@ -348,11 +348,17 @@ export default Ember.Mixin.create({
   */
   _generateSerializedHasMany(snapshot, relationship) {
     let hasMany = snapshot.hasMany(relationship.key);
-    return Ember.A(hasMany).map((embeddedSnapshot) => {
-      var embeddedJson = embeddedSnapshot.record.serialize({ includeId: true });
+    let manyArray = Ember.A(hasMany);
+    let ret = [];
```
Comment (stefanpenner) [inline addon/-private/serializers/embedded-records-mixin.js]: Let's pre alloc the array, looks like we know the size

File: addon/-private/serializers/embedded-records-mixin.js [hunk 2]
```diff
@@ -450,20 +456,22 @@ export default Ember.Mixin.create({
...
       hash.included = hash.included || [];
       hash.included.push(data);
       if (included) {
         hash.included.push(...included);
       }
-
-      return { id: data.id, type: data.type };
-    });
+      
+      hasMany.push({ id: data.id, type: data.type });
```
Comment (stefanpenner) [inline addon/-private/serializers/embedded-records-mixin.js]: ^^

File: addon/-private/serializers/json-api-serializer.js [hunk 3]
```diff
@@ -109,11 +109,25 @@ const JSONAPISerializer = JSONSerializer.extend({
     if (Ember.typeOf(documentHash.data) === 'object') {
       documentHash.data = this._normalizeResourceHelper(documentHash.data);
     } else if (Ember.typeOf(documentHash.data) === 'array') {
-      documentHash.data = documentHash.data.map(this._normalizeResourceHelper, this);
+      let ret = [];
```
Comment (stefanpenner) [inline addon/-private/serializers/json-api-serializer.js]: ^

File: addon/-private/serializers/json-api-serializer.js [hunk 4]
```diff
@@ -109,11 +109,25 @@ const JSONAPISerializer = JSONSerializer.extend({
...
+        let data = documentHash.data[i];
+        ret.push(this._normalizeResourceHelper(data));
+      }
+
+      documentHash.data = ret;
     }
 
     if (Ember.typeOf(documentHash.included) === 'array') {
-      documentHash.included = documentHash.included.map(this._normalizeResourceHelper, this);
+      let ret = [];
```
Comment (stefanpenner) [inline addon/-private/serializers/json-api-serializer.js]: Etc...

File: addon/-private/serializers/embedded-records-mixin.js (hunk 1 above)
Comment (chadhietala) [inline addon/-private/serializers/embedded-records-mixin.js]: Fixed.

File: addon/-private/serializers/embedded-records-mixin.js [hunk 5]
```diff
@@ -450,11 +456,13 @@ export default Ember.Mixin.create({
   */
   _extractEmbeddedHasMany(store, key, hash, relationshipMeta) {
     let relationshipHash = get(hash, `data.relationships.${key}.data`);
+    let hasMany = new Array(relationshipHash.length);
```
Comment (stefanpenner) [inline addon/-private/serializers/embedded-records-mixin.js]: maybe do this after the conditional on 460

File: addon/-private/serializers/json-api-serializer.js [hunk 6]
```diff
@@ -109,11 +109,25 @@ const JSONAPISerializer = JSONSerializer.extend({
...
+
+      for (let i = 0, l = documentHash.data.length; i < l; i++) {
+        let data = documentHash.data[i];
+        ret[i] = this._normalizeResourceHelper(data);
+      }
+
+      documentHash.data = ret;
     }
 
     if (Ember.typeOf(documentHash.included) === 'array') {
```
Comment (stefanpenner) [inline addon/-private/serializers/json-api-serializer.js]: i wonder if `Array.isArray(documentHash.data)` is sufficient (As it is much faster)

**Comment by chadhietala** [comment]
Windows failed because its windows. I didn't realized that some of the array methods were actually Maps being based around that conforms to the iterable interface.

File: addon/-private/system/store.js [hunk 7]
```diff
@@ -578,11 +578,13 @@ Store = Service.extend({
   */
   findByIds(modelName, ids) {
     assert('Passing classes to store methods has been removed. Please pass a dasherized string instead of '+ Ember.inspect(modelName), typeof modelName === 'string');
-    var store = this;
+    let promises = [];
```
Comment (stefanpenner) [inline addon/-private/system/store.js]: pre alloc

File: addon/-private/system/store.js [hunk 8]
```diff
@@ -841,7 +852,13 @@ Store = Service.extend({
     @return {Promise} promise
   */
   findMany(internalModels) {
-    return Promise.all(internalModels.map((internalModel) => this._findByInternalModel(internalModel)));
+    let finds = [];
```
Comment (stefanpenner) [inline addon/-private/system/store.js]: pre alloc

**Comment by topaxi** [comment]
Just curious, how much does this affect perf and gc pressure? :)

**Comment by chadhietala** [comment]
TL;DR Don't put things into new generation space to allow less GC events which must stop the world to collect.

@topaxi It depends on several factors but the largest is the granularity of your models and relationships between them. On mobile devices where you are much more constrained you need to be very observant of allocations. Contrary to popular belief engines like V8 create a "context" object when you create a closure, regardless of "functional is future" soup of the day :wink:. From my understanding*, these are new and short lived they are placed into new generation space which has a size limit. Once that limit is reached a minor GC scavenger runs to release the unreachable. This must stop the world to collect, so if you can perform the same amount of work with less allocations you should.

* This is my interpretation of this and it may be incorrect. For more information you might want to read this. https://strongloop.com/strongblog/node-js-performance-garbage-collection/

**Comment by topaxi** [comment]
@chadhietala thanks a lot for the detailed explanation!

**Comment by fivetanley** [comment]
Does this have benefits on runtimes that aren't V8?

**Comment by fivetanley** [comment]
It would also be great to see before/after numbers from this commit.

**Comment by krisselden** [comment]
Why would it not? All of them allocate for closures and not being careful with them is problematic.

**Comment by krisselden** [comment]
The is an excessive amount of looping and nested looping for relationship setup especially if you have not specified the inverse. The forEach and map closures likely could be made to avoid repeatedly creating them but in hot paths we've already demonstrated this is a hot path, profile an app that uses ember data, you will see relationshipsByName.

No, one single spot of waste is going to make a huge difference but we have to get rid of a lot of them to have an big impact.  When a low end android phone is about 10x slower it becomes a bigger deal.

File: addon/-private/system/model/errors.js [hunk 9]
```diff
@@ -272,14 +272,24 @@ export default Ember.ArrayProxy.extend(Ember.Evented, {
     @private
   */
   _findOrCreateMessages(attribute, messages) {
-    var errors = this.errorsFor(attribute);
+    let errors = this.errorsFor(attribute);
+    let _messages = [];
```
Comment (stefanpenner) [inline addon/-private/system/model/errors.js]: alloc

**Comment by fivetanley** [comment]
Often these kind of commits feel like insider baseball. It'd be great if we could measure these like we do in the [Ember Performance Suite](https://github.com/eviltrout/ember-performance/tree/master/app/templates). Chad's explanation was great and needed, and having documentation in the form of Pull Requests or commit messages gives us resources to come back to for those of us who aren't familiar with performance tuning.

File: addon/-private/system/model/errors.js (hunk 9 above)
Comment (fivetanley) [inline addon/-private/system/model/errors.js]: Are errors generated often enough for this to matter?

File: addon/-private/serializers/embedded-records-mixin.js [hunk 10]
```diff
@@ -348,11 +348,17 @@ export default Ember.Mixin.create({
   */
   _generateSerializedHasMany(snapshot, relationship) {
     let hasMany = snapshot.hasMany(relationship.key);
-    return Ember.A(hasMany).map((embeddedSnapshot) => {
-      var embeddedJson = embeddedSnapshot.record.serialize({ includeId: true });
+    let manyArray = Ember.A(hasMany);
+    let ret = new Array(manyArray.length);
+
+    for (let i = 0, l = manyArray.length; i < l; i++) {
```
Comment (stefanpenner) [inline addon/-private/serializers/embedded-records-mixin.js]: Caching the lengths isn't something need to do, lets not.

**Comment by stefanpenner** [comment]
@krisselden i suspect ^ will have measurable improvements, and even if there are none (unlikely) displaying the due diligence is important.

Doing so will also help others build a better mental model for this sort of optimization.

**Comment by runspired** [comment]
I'd love to see before/after data, cause to me this looks potentially like we're trading a small amount of execution time and more verbose code for a smaller chance of hitting a blocking GC.

**Comment by stefanpenner** [comment]
@runspired it's a known improvement. I'm merely asking for confirmation of the impact. If we don't see an improvement it is likely the result of some other issue. Which would be worth exploring further as it may uncover something we are not yet sure about.

The explicit code, is brain dead simple for the jit to handle, the callback variant is significantly more complex and as far as we know no runtime can yet reduce the existing code into anything nearly as performant and  as the patched variant should be. Maybe someday we will be blessed with more zero cost abstraction run times...


These loops are hit so often, and are themselves in other loops it's worth ensuring we aren't just trolling ourselves with some slip up

**Comment by chadhietala** [comment]
Will try to grab numbers tomorrow.

**Comment by chadhietala** [comment]
Probably some conflation going on here but appears to confirm that the hypothesis.

__Before:__
screen shot 2015-12-14 at 11 44 10 am

__After:__

screen shot 2015-12-14 at 11 41 47 am

**Comment by stefanpenner** [comment]
@fivetanley my above comment has not been addreseed https://github.com/emberjs/data/pull/3988#discussion_r47439430

**Comment by fivetanley** [comment]
We decided to merge this in the core team meeting. The Ember and Ember Data codebases both use this pattern of caching length, it doesn't seem enough to block merging it.

**Comment by stefanpenner** [comment]
@fivetanley can you address it then please, it would be unfortunate to continue to propogate this cargo culting
